### PR TITLE
[ENH] NiftiMasker: added time mask

### LINKED ISSUE #165: NiftiMasker should "time-mask" 4D volumes
In examples (e.g. plot_haxby_searchlight.py), we often do something like:

```python
fmri_img = nibabel.load(dataset_files.func)

### Restrict to faces and houses ################################
condition_mask = np.logical_or(conditions == 'face')
fmri_img = nibabel.Nifti1Image(fmri_img.get_data()[..., condition_mask],
                               fmri_img.get_affine().copy())

### Mask data ###############################################
mask_img = nibabel.load(dataset_files.mask)
nifti_masker = NiftiMasker(mask=mask_img)
fmri_masked = nifti_masker.fit_transform(fmri_img)
```

It would be simpler to provide `condition_mask` directly to the NiftiMasker to avoid creating a temporary image.

On top of making things easier to use, one important consideration is that some preprocessings performed by the NiftiMasker (such as detrending) might need the complete time serie to be accurate.
The NiftiMasker should then of course be able to choose whether it has to process the whole time serie or only the desired volumes according to the preprocessings that have to be performed. In the first case, is it important not to have masked the data before, while computation time would be saved in the second case (potentially a lot).

### PR DESCRIPTION
**Pull request by banilo**
tries to fix #165

### CODE REVIEW

**Comment by lesteve** [comment]
@banilo could you look at the merge conflicts on this one?

**Comment by banilo** [comment]
I promise to do it still some time today...

2015-01-14 10:36 GMT+01:00 Loïc Estève <notifications@github.com>:

> @banilo <https://github.com/banilo> could you look at the merge conflicts
> on this one?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/291#issuecomment-69890735>.
>



-- 
Viele Grüße,
Danilo

**Comment by banilo** [comment]
fixed merge conflicts + restored empty lines at file end + squashed

File: nilearn/input_data/base_masker.py [hunk 1]
```diff
@@ -38,6 +40,40 @@ def filter_and_mask(imgs, mask_img_,
 
     mask_img_ = _utils.check_niimg(mask_img_, ensure_3d=True)
     imgs = _utils.check_niimgs(imgs, accept_3d=True)
+    if mask_time is None:
+        imgs = _utils.check_niimgs(imgs, accept_3d=True)
+    else:
+        # check type and value of mask_time
+        if (not isinstance(mask_time, list) or
+            len(mask_time) == 0 or
+            not isinstance(mask_time[0], bool)):
```
Comment (lesteve) [inline nilearn/input_data/base_masker.py]: It feels quite restrictive that mask_time has to be a list of bool, shouldn't it be something that can be passed into a numpy array indexing operation:

```python
imgs = Nifti1Image(
    imgs.get_data()[..., np.asarray(mask_time)],
    imgs.get_affine())
```

For example, a slice object, a list of indices, etc...

Not sure what's the best way to make sure that mask_time can be used inside a indexing operation.

Comment (banilo) [inline nilearn/input_data/base_masker.py]: Why not, it's a design choice. Boolean arrays is however what naturally comes out of target/non-target label comparison operations.

Comment (GaelVaroquaux) [inline nilearn/input_data/base_masker.py]: Agreed with Loic. I wouldn't even call 'asarray' on it, so that it is
possible to pass Slice objects.

And do we want to call it 'mask_time' or 'time_mask' :)

Comment (banilo) [inline nilearn/input_data/base_masker.py]: > And do we want to call it 'mask_time' or 'time_mask' :)

I went for 'mask_time' in analogy to 'mask_img', to allude to the 4D-3D complementary. Should we change that?

Comment (banilo) [inline nilearn/input_data/base_masker.py]: > Agreed with Loic. I wouldn't even call 'asarray' on it, so that it is
> possible to pass Slice objects.

Can't we combine this with Loics ongoing indexing/slicing PR?

Comment (bthirion) [inline nilearn/input_data/base_masker.py]: Indeed, it is unclear for me what are the use cases addressed here that are not also addressed in the indexing/slicing PR ?
@banilo, do you have any example to motivate it ?

Comment (GaelVaroquaux) [inline nilearn/input_data/base_masker.py]: > Indeed, it is unclear for me what are the use cases addressed here that
> are not also addressed in the indexing/slicing PR ?

The indexing/slicing PR does make things much easier. However, suppose
that you want, eg to remove the 3 first volumes. With Danilo's PR, it is
trivial to do this in a pipeline, in distributed computing. Manually
using slicing would require more work.

However, the slicing PR should be merged, and then this PR updated to use
it.

Comment (banilo) [inline nilearn/input_data/base_masker.py]: A practical use case would be sub-selection as part of some sklearn-pipeline. Using this PR, memory is safed by avoiding creation of a tempory (sliced) nifti to pass on to the masker (index is provided at the time of masker instance creation). Just having idexing/slicing from Loic's PR cannot avoid this in a pipeline setting. ...I think.

Comment (bthirion) [inline nilearn/input_data/base_masker.py]: All right, thanks.

Comment (AlexandreAbraham) [inline nilearn/input_data/base_masker.py]: >  Just having idexing/slicing from Loic's PR cannot avoid this in a pipeline setting. ...I think.

You have the exact same behavior if you use Loïc's PR and a NiftiMasker (and do the loop yourself).

Comment (eickenberg) [inline nilearn/input_data/base_masker.py]: > Using this PR, memory is safed

line 74 makes a copy... 

If Loic's slicing PR permits slices into memmapped .nii files, then that is the best possible way of saving memory.

So the sole advantage of having time slicing (once Loic's is merged) in the NiftiMasker is the fact that it can go in a pipeline.

**Comment by AlexandreAbraham** [comment]
I have seen that time masking is done before the other pre-treatments. It makes sens the remove noisy scans at the beginning of a session, but for a condition mask (like Haxby), it would make more sense to me to apply it after the pretreatments (detrending, filtering...). So I'm wondering if it's a good idea to keep this name.

**Comment by banilo** [comment]
@AlexandreAbraham: Do you think it makes sense to include an option for either pre- or post-pretreatment application of the 4D filter?

**Comment by GaelVaroquaux** [comment]
> @AlexandreAbraham: Do you think it makes sense to include an option for either
> pre- or post-pretreatment application of the 4D filter?

We need to make a choice. Too many options are detrimental. It should
probably be 'pre application', because these are the only semantics that
make sens in general. However, this will blend a bit the masked data with
the non masked data in the case of time filtering, so we will have to
warn the user.

**Comment by banilo** [comment]
Ok, time masking now uses Loics index_img :)

File: nilearn/input_data/tests/test_nifti_masker.py [hunk 2]
```diff
@@ -10,6 +10,7 @@
 from tempfile import mkdtemp
 import shutil
 import os
+import numpy as np
```
Comment (lesteve) [inline nilearn/input_data/tests/test_nifti_masker.py]: import is already a few lines below.

File: nilearn/input_data/nifti_masker.py [hunk 3]
```diff
@@ -72,6 +72,14 @@ class NiftiMasker(BaseMasker, CacheMixin):
         to fine-tune mask computation. Please see the related documentation
         for details.
 
+    mask_time : Any type compatible with numpy array indexing
+        Mask for the niimgs along time/4D dimension. Thus complements
+        3D masking by mask_img argument. Data preprocessing will ignore
+        the time mask, whereas NiftiMasker.transform will only consider
+        time-masked images. Use case: Avoid creation of a temporary 4D data
+        nifti, potentially expensive in memory, before application of
+        NiftiMasker.
```
Comment (lesteve) [inline nilearn/input_data/nifti_masker.py]: > Data preprocessing will ignore the time mask whereas NiftiMasker.transform will only consider time-masked images.

I thought this wasn't true anymore, i.e. the indexing was done before the other treatments like filtering.

> Use case: Avoid creation of a temporary 4D data nifti, potentially expensive in memory, before application of NiftiMasker.

Not sure what you mean by that, it seems to be that indexing does create a temporary 4D image but maybe I am missing something.

Comment (lesteve) [inline nilearn/input_data/nifti_masker.py]: Also maybe the "pipelinability" should be mentioned in this docstring.

**Comment by lesteve** [comment]
1. I am still in favor of time_mask vs mask_time, be it only because it is closer to how you would call it in plain English.
2. Just curious, can some of the examples use this?

**Comment by GaelVaroquaux** [comment]
>  1. I am still in favor of time_mask vs mask_time, be it only because it is
>     closer to how you would call it in plain English.

+1

File: nilearn/input_data/tests/test_nifti_masker.py [hunk 4]
```diff
@@ -110,6 +111,37 @@ def test_mask_3d():
         assert_raises(TypeError, masker.fit)
 
 
+def test_mask_4d():
+    # Dummy mask
+    mask = np.zeros((10, 10, 10))
+    mask[3:7, 3:7, 3:7] = 1
+    mask_img = Nifti1Image(mask, np.eye(4))
+    n_mask_vox = len(np.where(mask == True)[0])
```
Comment (lesteve) [inline nilearn/input_data/tests/test_nifti_masker.py]: This is quite convoluted. I am sure there must be something better than this. mask.sum() should work but maybe there is something more generic.

File: nilearn/input_data/tests/test_nifti_masker.py [hunk 5]
```diff
@@ -110,6 +111,37 @@ def test_mask_3d():
...
+    masker5.fit()
+    data_trans = masker5.transform(data_imgs)
+    assert_true(np.all(data_trans))
+    assert_true((2, n_mask_vox) == data_trans.shape)  # verify mask application
+
+    masker5 = NiftiMasker(mask_img=mask_img, mask_time=mask_time)
+    masker5.fit()
+    data_trans = masker5.transform(data_img_4d)
+    assert_true(np.all(data_trans))
+    assert_true((2, n_mask_vox) == data_trans.shape)  # verify mask application
```
Comment (lesteve) [inline nilearn/input_data/tests/test_nifti_masker.py]: Could we not check that applying index_img and then the mask without time_mask is the same as applying the mask with time_mask to the original image?

**Comment by eickenberg** [comment]
+1

mask_time is weird

timecourse_mask would be another option.

however, the axis we are referring to can be more general than time. we
could also select conditions with it for decoding, see e.g. the haxby
example. in this perspective the name time_mask is a little restrictive.
samples_mask would be more generic, and still understandable/correct

File: nilearn/input_data/nifti_masker.py [hunk 6]
```diff
@@ -72,6 +72,14 @@ class NiftiMasker(BaseMasker, CacheMixin):
         to fine-tune mask computation. Please see the related documentation
         for details.
 
+    mask_time : Any type compatible with numpy array indexing
+        Mask for the niimgs along time/4D dimension. Thus complements
```
Comment (lesteve) [inline nilearn/input_data/nifti_masker.py]: minor comment, I would use `time/fourth dimension` rather than `time/4D dimension`

File: nilearn/input_data/tests/test_nifti_masker.py (hunk 5 above)
Comment (banilo) [inline nilearn/input_data/tests/test_nifti_masker.py]: Ok

File: nilearn/input_data/tests/test_nifti_masker.py (hunk 4 above)
Comment (banilo) [inline nilearn/input_data/tests/test_nifti_masker.py]: In real-world (float typed) binary mask niftis this did sometimes not give the same result, but here I will shift to sum() as suggested.

File: nilearn/input_data/nifti_masker.py (hunk 3 above)
Comment (banilo) [inline nilearn/input_data/nifti_masker.py]: I will make sure the 4D mask is 'pre application', as suggested by Gael

File: nilearn/input_data/tests/test_nifti_masker.py (hunk 4 above)
Comment (lesteve) [inline nilearn/input_data/tests/test_nifti_masker.py]: Depending on your use case, you could use something like `mask.astype(bool).sum()` or `np.isclose(mask, 0).sum()`. Note that the latter has some default absolute and relative tolerance that you may want to tweak.

**Comment by AlexandreAbraham** [comment]
> however, the axis we are referring to can be more general than time. we could also select conditions with it for decoding, see e.g. the haxby example. in this perspective the name time_mask is a little restrictive. samples_mask would be more generic, and still understandable/correct

In the case of Haxby example, don't you select conditions after the pre-treatment (ie detrending, etc)?

**Comment by eickenberg** [comment]
detrending and standardization cannot be done on the haxby example before
the chunks are cut up. One can also get by without doing it. In general I
was thinking more along the lines of series of beta maps from a rapid
event-related paradigm. You don't detrend or standardize those.

**Comment by GaelVaroquaux** [comment]
> detrending and standardization cannot be done on the haxby example before
> the chunks are cut up.

Why?

**Comment by eickenberg** [comment]
because they come from different runs. It doesn't make sense to detrend or
standardize over the borders of runs. scipy.signal.detrend takes
breakpoints, but they are not exposed in the niftimasker. I also don't
think that this is a general case. As already discussed with Bertrand,
datasets from nowadays are usually kept in one file per run because they
are already large on their own. It is only with datasets that at present
one can consider "old" that such concatenations become feasible.

**Comment by GaelVaroquaux** [comment]
> because they come from different runs.

One file of the Habxy dataset contains only one run. We abusively call
the "chunks" sessions, but they are not sessions in the usual sens.

**Comment by eickenberg** [comment]
They are 12 acquisition runs, the scanning is stopped after each one of them (that is my point) - but yes, one scanner session. Total acquisition time 3630 seconds.

**Comment by eickenberg** [comment]
(All in one file, runs labeled from 0 to 11 in the column "chunk")

**Comment by GaelVaroquaux** [comment]
> (All in one file, runs labeled from 0 to 11 in the column "chunk")

Fair enough. It seems that, to my surprise, you are right.

Anyhow, I believe that centering (though not standardizing or frequency
filtering) would be taken care off by defining the session vector as a
confound.

**Comment by eickenberg** [comment]
:) I had to reproduce the original study once, so I was forced to read all the small print at one point.

> centering would be taken care off by defining the session vector as a confound.

Indeed, in a GLM, but we have typically done decoding directly on the data.

My point is that one could use the "Time Masker" to mask conditions here. However, since these data are still temporally structured, the example is bad. If you consider a dataset consisting of many beta maps (such as an event-related paradigm), the is nothing time-related at all involved, but one may still be tempted to use the "Time Masking" to select images according to certain conditions.

Hence my proposition to use a name that is more general than the word "time", e.g. "sample". That said, I have no problem using something that has the name "time" in it on my non-temporal 4th axis.

**Comment by GaelVaroquaux** [comment]
>     centering would be taken care off by defining the session vector as a
>     confound.

> Indeed, in a GLM, but we have typically done decoding directly on the data.

The nifti maskers do take "confounds" as arguments to their transforms.

> Hence my proposition to use a name that is more general than the word "time",
> e.g. "sample". That said, I have no problem using something that has the name
> "time" in it on my non-temporal 4th axis.

I agree with that. "sample_mask" seems like a reasonnable name.

**Comment by banilo** [comment]
> "sample_mask" seems like a reasonnable name

+1 on this

**Comment by banilo** [comment]
Comments hopefully addressed. Mask is currently applied pre-treatment.

File: nilearn/input_data/nifti_masker.py [hunk 7]
```diff
@@ -72,6 +72,13 @@ class NiftiMasker(BaseMasker, CacheMixin):
         to fine-tune mask computation. Please see the related documentation
         for details.
 
+    sample_mask : Any type compatible with numpy array indexing
+        Mask for the niimgs along time/fourth dimension. Thus complements
+        3D masking by mask_img argument. Data preprocessing will ignore
+        the time mask, whereas NiftiMasker.transform will only consider
+        time-masked images. Use case: Data subselection as part of a
+        sklearn pipeline.
```
Comment (GaelVaroquaux) [inline nilearn/input_data/nifti_masker.py]: Rather than 'usecase: ...' I would say: 'this is useful to perform...'

**Comment by GaelVaroquaux** [comment]
LGTM: minor rephrasing to do, and a rebase necessary to address merge conflicts.

**Comment by banilo** [comment]
rephrased + rebased.

**Comment by banilo** [comment]
Docstring updated!

File: nilearn/input_data/tests/test_nifti_masker.py [hunk 8]
```diff
@@ -110,6 +111,39 @@ def test_mask_3d():
...
+
+    # Dummy data
+    data = np.zeros((10, 10, 10, 3))
+    data[..., 0] = 1
+    data[..., 1] = 2
+    data[..., 2] = 3
+    data_img_4d = Nifti1Image(data, np.eye(4))
+    data_imgs = [Nifti1Image(data[..., 0], np.eye(4)),
+        Nifti1Image(data[..., 1], np.eye(4)),
+        Nifti1Image(data[..., 2], np.eye(4))]
```
Comment (lesteve) [inline nilearn/input_data/tests/test_nifti_masker.py]: You may as well use index_img now that it exists.

Comment (banilo) [inline nilearn/input_data/tests/test_nifti_masker.py]: Should the tests not be as basic and transparent as possible?

File: nilearn/input_data/tests/test_nifti_masker.py [hunk 9]
```diff
@@ -110,6 +111,39 @@ def test_mask_3d():
...
+
+    # check whether transform is indeed selecting niimgs subset
+    sample_mask = (True, False, True)
+    masker = NiftiMasker(mask_img=mask_img, sample_mask=sample_mask)
+    masker.fit()
+    data_trans = masker.transform(data_imgs)
+    data_trans_direct = data_img_4d.get_data()[mask_bool, :][..., sample_mask]
+    data_trans_direct = np.swapaxes(data_trans_direct, 0, 1)
```
Comment (lesteve) [inline nilearn/input_data/tests/test_nifti_masker.py]: Isn't it simpler to do (please use better names than mine):

```python

masker2 = NiftiMasker(mask_img=mask_img)
masker2.fit()
data_imgs_subset = index_img(data_imgs, sample_mask)
data_trans_direct = masker2.transform(data_imgs_subset)
```

Basically you want to make sure that using sample_mask in the masker is the same as using index_img on the data and then a masker without sample_mask.

File: nilearn/input_data/base_masker.py [hunk 10]
```diff
@@ -167,13 +174,14 @@ def transform_single_imgs(self, imgs, confounds=None, copy=True):
...
-                                verbose=self.verbose,
-                                confounds=confounds,
-                                copy=copy
+                              imgs, self.mask_img_,
+                              params,
+                              memory_level=self.memory_level,
+                              memory=self.memory,
+                              verbose=self.verbose,
+                              confounds=confounds,
+                              copy=copy,
```
Comment (lesteve) [inline nilearn/input_data/base_masker.py]: it looks like you moved away from PEP8 here

File: nilearn/input_data/tests/test_nifti_masker.py (hunk 8 above)
Comment (lesteve) [inline nilearn/input_data/tests/test_nifti_masker.py]: Yes. By the way, just curious, any particular reason you are not using data_img_4d in the transform call?

File: nilearn/input_data/base_masker.py [hunk 11]
```diff
@@ -37,7 +39,11 @@ def filter_and_mask(imgs, mask_img_,
         class_name = enclosing_scope_name(stack_level=2)
 
     mask_img_ = _utils.check_niimg(mask_img_, ensure_3d=True)
-    imgs = _utils.check_niimgs(imgs, accept_3d=True)
+
+    if sample_mask is not None:
+        imgs = image.index_img(imgs, sample_mask)
+    else:
+        imgs = _utils.check_niimgs(imgs, accept_3d=True)
```
Comment (lesteve) [inline nilearn/input_data/base_masker.py]: Maybe it's just me but I would do:

```python
imgs = _utils.check_niimgs(imgs, accept_3d=True)

if sample_mask is not None:
    imgs = image.index_img(imgs, sample_mask)
```

File: nilearn/input_data/tests/test_nifti_masker.py (hunk 8 above)
Comment (banilo) [inline nilearn/input_data/tests/test_nifti_masker.py]: That is used in line 143.

File: nilearn/input_data/base_masker.py (hunk 11 above)
Comment (banilo) [inline nilearn/input_data/base_masker.py]: It has been like that before, Alex wanted me to change it this way.

**Comment by banilo** [comment]
Ok, done.

File: nilearn/input_data/base_masker.py (hunk 11 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/base_masker.py]: I suggested that because I thought that `index_img` could be faster if `imgs` was a list of 3D imgs. Given that it's not the case, I would be in favor of what @lesteve proposes.

Comment (lesteve) [inline nilearn/input_data/base_masker.py]: OK just this one to do and it should be good to go.

Comment (banilo) [inline nilearn/input_data/base_masker.py]: changed back to how it was before, as suggested.

**Comment by lesteve** [comment]
Looks good, merging, thanks a lot.
